### ISSUE TITLE
Extending the Directive class

### ISSUE THREAD

**Issue by flashbackzoo** [nswdpc/silverstripe-csp#6]
A project I'm working needs to add `unsafe-eval` to a directive. Adding the "Unsafe eval" checkbox is simple enough using a `DataExtension` but the `Directive` class doesn't provide an extension point to update the actual directive value.

As a workaround I've replaced the `Directive` class with a custom subclass, using the dependency injector, and overridden the `getDirectiveValue` method with what I need.

So I guess this is a feature request. What are your thoughts on adding an extension point to the end of `Directive::getDirectiveValue` e.g.

```
...

$this->extend('updateDirectiveValue', $value);

return $value;
```

**Comment by JamesDPC** [comment]
Hi David

You should be able to add that in the Rules & Restrictions multi value field (it has to be quoted).
When I added the below, the system produced this header:

```
content-security-policy-report-only: script-src 'self' 'unsafe-eval';
```

![image](https://user-images.githubusercontent.com/69664712/174689983-eed2be71-0487-4962-bfbd-830791306049.png)



Does this work for you ?

Cheers
James

**Comment by JamesDPC** [comment]
I should add I'd deliberately avoided adding an unsafe eval checkbox to the interface. An author who adds that manually is assumed to be across eval badness. There is a case to have unsafe-inline handled this way, as well.

**Comment by flashbackzoo** [comment]
Yeah that works, thanks @JamesDPC 🤦 😄

**Comment by JamesDPC** [comment]
No worries, if it's a case of the "Rules and Restrictions" multi value field not being obvious we can improve that. Maybe it should be renamed to "Extra Directive values" 🤔

**Comment by flashbackzoo** [comment]
Maybe just an example or two added to the existing [rules and restrictions docs](https://github.com/nswdpc/silverstripe-csp/blob/master/docs/en/00_index.md#rules-and-restrictions)? Something like:

| Directive name | Comment |
| - | - |
| *.ytimg.com | YouTube static assets |
| 'unsafe-eval' | Allow Vue3 templates to compile at runtime |

**Comment by JamesDPC** [comment]
@flashbackzoo  see that commit ^ just added for some changes.

It includes some examples shown in the directive editing area:

![image](https://user-images.githubusercontent.com/69664712/207789456-6a68c39a-52cf-40f3-8bb4-b0085200fbbf.png)

I think this meets the requirements, so I'll merge that one in and tag a release.

**Comment by JamesDPC** [comment]
Resolved in [v0.3.1](https://github.com/nswdpc/silverstripe-csp/releases/tag/v0.3.1)
